ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java

```diff
@@ -435,55 +187,71 @@ private void initObjectsForMetrics() throws Exception {
-
-    deltaCache = CacheBuilder.newBuilder()
-      .expireAfterWrite(duration, TimeUnit.SECONDS)
-      .removalListener(notification -> removalPredicate(deltaTopN, notification))
-      .softValues()
-      .build();
-
-    smallDeltaCache = CacheBuilder.newBuilder()
-      .expireAfterWrite(duration, TimeUnit.SECONDS)
-      .removalListener(notification -> removalPredicate(smallDeltaTopN, notification))
-      .softValues()
-      .build();
-
-    obsoleteDeltaCache = CacheBuilder.newBuilder()
-      .expireAfterWrite(duration, TimeUnit.SECONDS)
-      .removalListener(notification -> removalPredicate(obsoleteDeltaTopN, notification))
-      .softValues()
-      .build();
-  }
-
-  private static Comparator<Pair<String, Integer>> getComparator() {
-    return Comparator.comparing(Pair::getValue);
-  }
+  private final class ReportingTask implements Runnable {
 
-  private void removalPredicate(BlockingQueue<Pair<String, Integer>> topN, RemovalNotification notification) {
-    topN.removeIf(item -> item.getKey().equals(notification.getKey()));
-  }
+    private final TxnStore txnHandler;
 
-  private final class ReportingTask implements Runnable {
+    private ReportingTask(TxnStore txnHandler) {
+      this.txnHandler = txnHandler;
+    }
     @Override
     public void run() {
       Metrics metrics = MetricsFactory.getInstance();
       if (metrics != null) {
-        obsoleteDeltaCache.cleanUp();
-        obsoleteDeltaObject.updateAll(obsoleteDeltaCache.asMap());
+        try {
+          LOG.debug("Called reporting task.");
+          List<CompactionMetricsData> deltas = txnHandler.getTopCompactionMetricsDataPerType(maxCacheSize);
+          Map<String, Integer> deltasMap = deltas.stream()
+              .filter(d -> d.getMetricType() == CompactionMetricsData.MetricType.NUM_DELTAS).collect(
+              Collectors.toMap(item -> getDeltaCountKey(item.getDbName(), item.getTblName(), item.getPartitionName()),
+                  CompactionMetricsData::getMetricValue));
+          deltaObject.updateAll(deltasMap);
+
+          Map<String, Integer> smallDeltasMap = deltas.stream()
+              .filter(d -> d.getMetricType() == CompactionMetricsData.MetricType.NUM_SMALL_DELTAS).collect(
+              Collectors.toMap(item -> getDeltaCountKey(item.getDbName(), item.getTblName(), item.getPartitionName()),
+                  CompactionMetricsData::getMetricValue));
+          smallDeltaObject.updateAll(smallDeltasMap);
+
+          Map<String, Integer> obsoleteDeltasMap = deltas.stream()
+              .filter(d -> d.getMetricType() == CompactionMetricsData.MetricType.NUM_OBSOLETE_DELTAS).collect(
+              Collectors.toMap(item -> getDeltaCountKey(item.getDbName(), item.getTblName(), item.getPartitionName()),
+                  CompactionMetricsData::getMetricValue));
+          obsoleteDeltaObject.updateAll(obsoleteDeltasMap);
+        } catch (MetaException e) {
```

Comment:
Maybe catch all Throwables here just in case? (and also in run())

Reply:
Right, I will do that.